common/sodcrypto.cc

```diff
@@ -65,20 +64,113 @@ std::string sodDecryptSym(const std::string& msg, const std::string& key, Sodium
   }
 }
 #else
-std::string sodEncryptSym(const std::string& msg, const std::string& key, SodiumNonce& nonce)
+
+#define POLY1305_BLOCK_SIZE 16
+
+string newKey()
 {
-  return msg;
+  unsigned char key[CHACHA20_POLY1305_KEY_SIZE];
+  if (!RAND_priv_bytes(key, sizeof key)) {
+    throw std::runtime_error(
+        "Could not initialize random number generator for cryptographic functions - this is not recoverable");
+  }
+  return "\"" + Base64Encode(string((char*) key, sizeof key)) + "\"";
 }
-std::string sodDecryptSym(const std::string& msg, const std::string& key, SodiumNonce& nonce)
+
+
+std::string sodEncryptSym(const std::string& msg, const std::string& key, SodiumNonce& nonce)
 {
-  return msg;
+  unsigned char ciphertext[msg.length() + POLY1305_BLOCK_SIZE];
+  int len;
+  int ciphertext_len;
+  // Each thread gets its own cipher context
+  thread_local EVP_CIPHER_CTX *ctx = nullptr;
```

Comment:
Just thought about this, don't you want this to be `static thread_local` instead? Otherwise you are not reusing the context.
```suggestion
  static thread_local EVP_CIPHER_CTX *ctx = nullptr;
```

Reply:
Haha yes good point, I fully intended these to be static but didn’t type it for some reason!